### PR TITLE
Add ComponentHolderServerPacket to some packets

### PR DESCRIPTION
**Pull request by Marfien**
Added the `ComponentHolderServerPacket` interface to some packets for global translation if `MinestomAdventure.AUTOMATIC_COMPONENT_TRANSLATION` is enabled

### CODE REVIEW

File: src/main/java/net/minestom/server/event/instance/InstanceRegisteredEvent.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package net.minestom.server.event.instance;
+
+import net.minestom.server.event.trait.InstanceEvent;
+import net.minestom.server.instance.Instance;
+import org.jetbrains.annotations.NotNull;
+
+public class InstanceRegisteredEvent implements InstanceEvent {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/event/instance/InstanceRegisteredEvent.java:7]: This is not the right PR to introduce a new event

File: src/main/java/net/minestom/server/network/packet/server/play/ActionBarPacket.java [hunk 2]
```diff
@@ -21,4 +26,14 @@ public void write(@NotNull BinaryWriter writer) {
     public int getId() {
         return ServerPacketIdentifier.ACTION_BAR;
     }
+
+    @Override
+    public @NotNull Collection<Component> components() {
+        return Collections.singleton(this.text);
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/ActionBarPacket.java:32]: `Collections.singleton` should be replaced to `List.of` (we currently use `singleton` in some of our types, but this is a mistake)

File: src/main/java/net/minestom/server/network/packet/server/play/AdvancementsPacket.java [hunk 3]
```diff
@@ -41,10 +45,35 @@ public int getId() {
...
+        final var displayData = this.advancementMappings.stream().map(AdvancementMapping::value).map(Advancement::displayData).filter(Objects::nonNull).toList();
+        final var titles = displayData.stream().map(DisplayData::title).toList();
+        final var descriptions = displayData.stream().map(DisplayData::description).toList();
+
+        final var list = new ArrayList<Component>();
+
+        list.addAll(titles);
+        list.addAll(descriptions);
+
+        return list;
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/AdvancementsPacket.java:60]: You should return immutable lists

File: src/main/java/net/minestom/server/network/packet/server/play/AdvancementsPacket.java [hunk 4]
```diff
@@ -80,6 +119,16 @@ public void write(@NotNull BinaryWriter writer) {
             writer.writeVarIntList(criteria, BinaryWriter::writeSizedString);
             writer.writeVarIntList(requirements, BinaryWriter::write);
         }
+
+        @Override
+        public @NotNull Collection<Component> components() {
+            return this.displayData != null ? this.displayData.components() : Collections.emptyList();
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/AdvancementsPacket.java:125]: `List.of`

File: src/main/java/net/minestom/server/network/packet/server/play/BossBarPacket.java [hunk 5]
```diff
@@ -32,13 +37,27 @@ public void write(@NotNull BinaryWriter writer) {
         writer.write(action);
     }
 
+    @Override
+    public @NotNull Collection<Component> components() {
+        return this.action instanceof ComponentHolder<?> holder
+                ? holder.components()
+                : Collections.emptyList();
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/BossBarPacket.java:44]: `List.of`

File: src/main/java/net/minestom/server/network/packet/server/play/WindowItemsPacket.java [hunk 6]
```diff
@@ -32,4 +39,22 @@ public void write(@NotNull BinaryWriter writer) {
     public int getId() {
         return ServerPacketIdentifier.WINDOW_ITEMS;
     }
+
+    @Override
+    public @NotNull Collection<Component> components() {
+        final var list = new ArrayList<>(this.items);
+        list.add(this.carriedItem);
+
+        return list.stream().map(ItemStack::getDisplayName).filter(Objects::nonNull).collect(Collectors.toList());
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/WindowItemsPacket.java:48]: There is absolutely no reason to use `collect(Collectors.toList())` over `toList()` from JDK 17

File: src/main/java/net/minestom/server/utils/PacketUtils.java [hunk 7]
```diff
@@ -101,12 +102,24 @@ public static void sendPacket(@NotNull Audience audience, @NotNull ServerPacket
...
 
+    /**
+     * Checks if the {@link ServerPacket} is suitable to be wrapped into a {@link CachedPacket}.
+     * Note: {@link ComponentHoldingServerPacket}s are not translated inside a {@link CachedPacket}.
+     *
+     * @see CachedPacket#body()
+     * @see PlayerSocketConnection#writePacketSync(SendablePacket, boolean)
+     */
+    private static boolean shouldUseCachePacket(final @NotNull ServerPacket packet) {
+        return (!MinestomAdventure.AUTOMATIC_COMPONENT_TRANSLATION || !(packet instanceof ComponentHoldingServerPacket)) && GROUPED_PACKET;
```
Comment (TheMode) [inline src/main/java/net/minestom/server/utils/PacketUtils.java:120]: May be worth checking if the `ComponentHoldingServerPacket` has any component to actually translate. For example in packets containing items where you may not have to do anything

**Review by TheMode — CHANGES_REQUESTED** [review]

File: src/main/java/net/minestom/server/event/instance/InstanceRegisteredEvent.java (hunk 1 above)
Comment (Marfien) [reply src/main/java/net/minestom/server/event/instance/InstanceRegisteredEvent.java:7]: Oh, sorry
My bad

File: src/main/java/net/minestom/server/network/packet/server/play/ActionBarPacket.java (hunk 2 above)
Comment (Marfien) [reply src/main/java/net/minestom/server/network/packet/server/play/ActionBarPacket.java:32]: done

File: src/main/java/net/minestom/server/utils/PacketUtils.java (hunk 7 above)
Comment (Marfien) [reply src/main/java/net/minestom/server/utils/PacketUtils.java:120]: I'm not sure, if the way I implemented it is the best.

File: src/main/java/net/minestom/server/adventure/ComponentHolder.java [hunk 8]
```diff
@@ -40,4 +41,16 @@ default void visitComponents(@NotNull Consumer<Component> visitor) {
             visitor.accept(component);
         }
     }
+
+    default boolean hasTranslatableComponents() {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/adventure/ComponentHolder.java:45]: No new API. You should move it to `PacketUtils`

You should also be aware that this method may be incredibly hot. So worth optimizing (as I see multiple expensive streams)

File: src/main/java/net/minestom/server/network/packet/server/play/BossBarPacket.java [hunk 9]
```diff
@@ -63,6 +83,16 @@ public void write(@NotNull BinaryWriter writer) {
         public int id() {
             return 0;
         }
+
+        @Override
+        public @NotNull Collection<Component> components() {
+            return Collections.singleton(this.title);
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/BossBarPacket.java:89]: `List.of`

File: src/main/java/net/minestom/server/network/packet/server/play/ResourcePackSendPacket.java [hunk 10]
```diff
@@ -37,4 +42,14 @@ public void write(@NotNull BinaryWriter writer) {
     public int getId() {
         return ServerPacketIdentifier.RESOURCE_PACK_SEND;
     }
+
+    @Override
+    public @NotNull Collection<Component> components() {
+        return Collections.singleton(this.prompt);
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/ResourcePackSendPacket.java:48]: `List.of`

File: src/main/java/net/minestom/server/network/packet/server/play/SetSlotPacket.java [hunk 11]
```diff
@@ -27,6 +33,18 @@ public int getId() {
         return ServerPacketIdentifier.SET_SLOT;
     }
 
+    @Override
+    public @NotNull Collection<Component> components() {
+        final var displayname = this.itemStack.getDisplayName();
+
+        return displayname == null ? Collections.emptyList() : Collections.singleton(displayname);
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/SetSlotPacket.java:40]: `List.of`

File: src/main/java/net/minestom/server/network/packet/server/play/SetTitleSubTitlePacket.java [hunk 12]
```diff
@@ -21,4 +26,14 @@ public void write(@NotNull BinaryWriter writer) {
     public int getId() {
         return ServerPacketIdentifier.SET_TITLE_SUBTITLE;
     }
+
+    @Override
+    public @NotNull Collection<Component> components() {
+        return Collections.singleton(this.subtitle);
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/SetTitleSubTitlePacket.java:32]: `List.of`

File: src/main/java/net/minestom/server/network/packet/server/play/SetTitleTextPacket.java [hunk 13]
```diff
@@ -21,4 +26,14 @@ public void write(@NotNull BinaryWriter writer) {
     public int getId() {
         return ServerPacketIdentifier.SET_TITLE_TEXT;
     }
+
+    @Override
+    public @NotNull Collection<Component> components() {
+        return Collections.singleton(this.title);
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/SetTitleTextPacket.java:32]: `List.of`

File: src/main/java/net/minestom/server/network/packet/server/play/TeamsPacket.java [hunk 14]
```diff
@@ -36,6 +40,21 @@ public void write(@NotNull BinaryWriter writer) {
         writer.write(action);
     }
 
+    @Override
+    public @NotNull Collection<Component> components() {
+        return this.action instanceof ComponentHolder<?> holder ? holder.components() : Collections.emptyList();
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/TeamsPacket.java:45]: `List.of`

File: src/main/java/net/minestom/server/network/packet/server/play/WindowItemsPacket.java [hunk 15]
```diff
@@ -32,4 +39,22 @@ public void write(@NotNull BinaryWriter writer) {
...
+
+        return list.stream().map(ItemStack::getDisplayName).filter(Objects::nonNull).toList();
+    }
+
+    @Override
+    public @NotNull ServerPacket copyWithOperator(@NotNull UnaryOperator<Component> operator) {
+        return new WindowItemsPacket(
+                this.windowId,
+                this.stateId,
+                this.items.stream().map(stack -> stack.withDisplayName(operator)).collect(Collectors.toList()),
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/WindowItemsPacket.java:56]: `toList`

**Review by TheMode — CHANGES_REQUESTED** [review]
`List.of` shall be used instead of the legacy `Collections` methods.

Additionally a test for adventure translation would be great (sending a packet and ensuring that the result is properly translated). You can use our testing framework

**Pull request merged**